js/src/components/tree-select-control/index.js

```diff
@@ -217,6 +235,93 @@ const TreeSelectControl = ( {
 		onChange( [ ...tags.map( ( el ) => el.id ) ] );
 	};
 
+	/**
+	 * Prepares and sets the search filter.
+	 * Filters of less than 3 characters are not considered, so we convert them to ''
+	 *
+	 * @param {Event} e Event returned by the On Change function
+	 */
+	const handleOnSearch = ( e ) => {
+		const search = e.target.value.trim();
+		setFilter( search.length >= 3 ? search : '' );
+	};
+
+	/**
+	 * Applies the strong tag in the Option label characters matching the filter
+	 *
+	 * @param {string} optionLabel The label to highlight
+	 * @param {number} matchPosition The position when the highlight starts based on the filter match
+	 * @return {JSX.Element|*} A highlighted label
+	 */
+	const highlightOptionLabel = ( optionLabel, matchPosition ) => {
+		if ( matchPosition < 0 || ! filter?.length ) return optionLabel;
+
+		return (
+			<span>
+				<span>{ optionLabel.substring( 0, matchPosition ) }</span>
+				<strong>
+					{ optionLabel.substring(
+						matchPosition,
+						matchPosition + filter.length
+					) }
+				</strong>
+				<span>
+					{ optionLabel.substring( matchPosition + filter.length ) }
+				</span>
+			</span>
+		);
+	};
+
+	/**
+	 * Filters an option and transforms its label if so
+	 *
+	 * This function mutates the option
+	 *
+	 * @param {Option} option The option to filter
+	 * @return {boolean} True if it pass the filter, false otherwise
+	 */
+	const filterOption = ( option ) => {
+		if ( hasChildren( option ) ) {
+			option.children = option.children.filter( filterOption );
+			return !! option.children.length;
```

Comment:
❓ 
By the design in Figma, parent options will still be displayed even if there are no matching children. Has the requirement changed?

![image](https://user-images.githubusercontent.com/17420811/158356992-cf135142-e1fb-46bf-b619-2c8baf4942c9.png)

Reply:
I have huge doubts about this as well. 

My main wonder is. If we show the continents, what happens if the user expand them? I guess if so, we should remove the chevron. 

I'll ask around in Component P2

Reply:
Another question (maybe for the future) is if the tree is multilevel, and we don't hide the parents. A parent without results but for example with 4-5 levels would be expandable several times to show nothing at the end.